File: backend-v1/pipeline/35-extract-form-features.js

'use strict';

/**
 * ============================================================
 * ZOKASCORE V2 — PIPELINE 35
 * FORM & H2H FEATURE EXTRACTION
 * ============================================================
 *
 * Purpose:
 *   Transform chronological ELO history into richer
 *   pre-match football intelligence features.
 *
 * IMPORTANT:
 *   Every feature is calculated BEFORE the current match.
 *
 *   The current match result is added to state only AFTER
 *   its feature row has been generated.
 *
 * This prevents future-data leakage.
 *
 * Output:
 *   data/ml/features_v2.csv
 *
 * Features:
 *
 *   ELO
 *   - home_elo_pre
 *   - away_elo_pre
 *   - elo_diff
 *
 *   Overall recent form
 *   - home_form_pts
 *   - away_form_pts
 *
 *   Venue-specific form
 *   - home_home_pts
 *   - away_away_pts
 *
 *   Recent goals
 *   - home_gf_avg
 *   - away_gf_avg
 *   - home_ga_avg
 *   - away_ga_avg
 *
 *   Head-to-head
 *   - h2h_hw_rate
 *   - h2h_d_rate
 *   - h2h_aw_rate
 *   - h2h_matches
 *
 * Target:
 *   HOME_WIN
 *   DRAW
 *   AWAY_WIN
 * ============================================================
 */

const fs = require('fs');
const path = require('path');

const ROOT = path.join(__dirname, '..');

const ELO_INDEX_FILE = path.join(
  ROOT,
  'data',
  'elo',
  'elo_processed_matches.json'
);

const OUTPUT_DIR = path.join(
  ROOT,
  'data',
  'ml'
);

const OUTPUT_FILE = path.join(
  OUTPUT_DIR,
  'features_v2.csv'
);

/**
 * ============================================================
 * HELPERS
 * ============================================================
 */

function readJson(file) {
  return JSON.parse(
    fs.readFileSync(file, 'utf8')
  );
}

/**
 * Keep only the most recent records needed for feature
 * extraction. This prevents unnecessary memory growth.
 */
function pushRecent(map, teamId, record) {
  if (!map.has(teamId)) {
    map.set(teamId, []);
  }

  const history = map.get(teamId);

  history.push(record);

  /*
   * We only need the recent five matches for this pipeline.
   * Keep a small safety buffer because venue filtering may
   * require looking through recent records.
   */
  if (history.length > 20) {
    history.splice(0, history.length - 20);
  }
}

/**
 * ============================================================
 * RECENT FORM
 * ============================================================
 */

function getRecentStats(
  teamRecent,
  teamId,
  venue = null
) {
  const history =
    teamRecent.get(teamId) || [];

  const relevant = venue
    ? history.filter(
        match => match.venue === venue
      )
    : history;

  const last5 = relevant.slice(-5);

  /*
   * No previous information.
   *
   * Neutral prior prevents NaN values while making sure
   * unknown history does not accidentally become a strong
   * positive or negative signal.
   */
  if (last5.length === 0) {
    return {
      matches: 0,
      pts: 0,
      gf_avg: 1.0,
      ga_avg: 1.0
    };
  }

  let pts = 0;
  let gf = 0;
  let ga = 0;

  for (const match of last5) {
    pts += Number(match.pts) || 0;
    gf += Number(match.gf) || 0;
    ga += Number(match.ga) || 0;
  }

  return {
    matches: last5.length,
    pts,
    gf_avg: gf / last5.length,
    ga_avg: ga / last5.length
  };
}

/**
 * ============================================================
 * H2H
 * ============================================================
 *
 * Store matchup independently of venue orientation.
 *
 * Example:
 *
 *   A vs B
 *   B vs A
 *
 * Both belong to:
 *
 *   A|B
 *
 * But each result is stored from the perspective of the
 * canonical first team.
 * ============================================================
 */

function getH2HKey(teamA, teamB) {
  return [
    String(teamA),
    String(teamB)
  ].sort().join('|');
}

function getH2HStats(
  h2hState,
  homeId,
  awayId
) {
  const key = getH2HKey(
    homeId,
    awayId
  );

  const state =
    h2hState.get(key) || {
      teamA: key.split('|')[0],
      teamB: key.split('|')[1],
      teamA_wins: 0,
      draws: 0,
      teamB_wins: 0
    };

  const total =
    state.teamA_wins +
    state.draws +
    state.teamB_wins;

  if (total === 0) {
    return {
      hw_rate: 0.0,
      d_rate: 0.0,
      aw_rate: 0.0,
      matches: 0
    };
  }

  const homeIsTeamA =
    String(homeId) === state.teamA;

  return {
    hw_rate: homeIsTeamA
      ? state.teamA_wins / total
      : state.teamB_wins / total,

    d_rate:
      state.draws / total,

    aw_rate: homeIsTeamA
      ? state.teamB_wins / total
      : state.teamA_wins / total,

    matches: total
  };
}

/**
 * ============================================================
 * UPDATE H2H STATE
 * ============================================================
 */

function updateH2H(
  h2hState,
  homeId,
  awayId,
  result
) {
  const key = getH2HKey(
    homeId,
    awayId
  );

  if (!h2hState.has(key)) {
    h2hState.set(key, {
      teamA: key.split('|')[0],
      teamB: key.split('|')[1],
      teamA_wins: 0,
      draws: 0,
      teamB_wins: 0
    });
  }

  const state = h2hState.get(key);

  const homeIsTeamA =
    String(homeId) === state.teamA;

  if (result === 'DRAW') {
    state.draws++;
    return;
  }

  const homeWon =
    result === 'HOME_WIN';

  const teamAWon =
    homeIsTeamA
      ? homeWon
      : !homeWon;

  if (teamAWon) {
    state.teamA_wins++;
  } else {
    state.teamB_wins++;
  }
}

/**
 * ============================================================
 * MAIN
 * ============================================================
 */

async function main() {
  console.log(
    '⚽ ZOKASCORE V2 — Pipeline 35: Form & H2H Feature Extraction'
  );

  console.log(
    '============================================================\n'
  );

  /**
   * ----------------------------------------------------------
   * Validate inputs
   * ----------------------------------------------------------
   */

  if (!fs.existsSync(ELO_INDEX_FILE)) {
    throw new Error(
      'ELO index not found. Run Pipeline 32 first.'
    );
  }

  fs.mkdirSync(
    OUTPUT_DIR,
    { recursive: true }
  );

  /**
   * ----------------------------------------------------------
   * Load ELO processed matches
   * ----------------------------------------------------------
   */

  console.log(
    '📚 Loading ELO processed index...'
  );

  const eloIndex =
    readJson(ELO_INDEX_FILE);

  let matches =
    Object.values(eloIndex);

  /**
   * ----------------------------------------------------------
   * Chronological ordering
   * ----------------------------------------------------------
   *
   * This is essential.
   *
   * Feature state must only contain information from
   * matches that happened BEFORE the current match.
   */

  matches.sort((a, b) => {
    const tA = a.date
      ? Date.parse(a.date)
      : Number.MAX_SAFE_INTEGER;

    const tB = b.date
      ? Date.parse(b.date)
      : Number.MAX_SAFE_INTEGER;

    return tA - tB;
  });

  console.log(
    `   ✅ Loaded ${matches.length.toLocaleString()} chronological matches.\n`
  );

  /**
   * ----------------------------------------------------------
   * State
   * ----------------------------------------------------------
   */

  const teamRecent = new Map();

  const h2hState = new Map();

  /**
   * ----------------------------------------------------------
   * CSV HEADER
   * ----------------------------------------------------------
   */

  const csvLines = [
    [
      'match_id',
      'date',
      'home_team_id',
      'away_team_id',

      'home_elo_pre',
      'away_elo_pre',
      'elo_diff',

      'home_form_pts',
      'away_form_pts',

      'home_home_pts',
      'away_away_pts',

      'home_gf_avg',
      'away_gf_avg',

      'home_ga_avg',
      'away_ga_avg',

      'h2h_hw_rate',
      'h2h_d_rate',
      'h2h_aw_rate',
      'h2h_matches',

      'target'
    ].join(',')
  ];

  /**
   * ----------------------------------------------------------
   * PROCESS
   * ----------------------------------------------------------
   */

  let processed = 0;
  let skipped = 0;

  for (const m of matches) {

    /**
     * --------------------------------------------------------
     * Validate match
     * --------------------------------------------------------
     */

    if (
      !m ||
      m.home_team_id == null ||
      m.away_team_id == null ||
      m.home_goals == null ||
      m.away_goals == null ||
      !m.result
    ) {
      skipped++;
      continue;
    }

    const homeId =
      String(m.home_team_id);

    const awayId =
      String(m.away_team_id);

    /**
     * --------------------------------------------------------
     * PRE-MATCH FORM
     * --------------------------------------------------------
     *
     * IMPORTANT:
     *
     * These are calculated BEFORE the current match is
     * inserted into state.
     */

    const homeOverall =
      getRecentStats(
        teamRecent,
        homeId,
        null
      );

    const awayOverall =
      getRecentStats(
        teamRecent,
        awayId,
        null
      );

    const homeHome =
      getRecentStats(
        teamRecent,
        homeId,
        'home'
      );

    const awayAway =
      getRecentStats(
        teamRecent,
        awayId,
        'away'
      );

    /**
     * --------------------------------------------------------
     * PRE-MATCH H2H
     * --------------------------------------------------------
     */

    const h2h =
      getH2HStats(
        h2hState,
        homeId,
        awayId
      );

    /**
     * --------------------------------------------------------
     * ELO
     * --------------------------------------------------------
     */

    const homeElo =
      Number(m.home_elo_before);

    const awayElo =
      Number(m.away_elo_before);

    if (
      !Number.isFinite(homeElo) ||
      !Number.isFinite(awayElo)
    ) {
      skipped++;
      continue;
    }

    const eloDiff =
      homeElo - awayElo;

    /**
     * --------------------------------------------------------
     * DATE
     * --------------------------------------------------------
     */

    const cleanDate =
      m.date
        ? String(m.date).split('T')[0]
        : '';

    /**
     * --------------------------------------------------------
     * CSV ROW
     * --------------------------------------------------------
     */

    csvLines.push([
      m.match_id,
      cleanDate,

      homeId,
      awayId,

      homeElo.toFixed(2),
      awayElo.toFixed(2),
      eloDiff.toFixed(2),

      homeOverall.pts,
      awayOverall.pts,

      homeHome.pts,
      awayAway.pts,

      homeOverall.gf_avg.toFixed(2),
      awayOverall.gf_avg.toFixed(2),

      homeOverall.ga_avg.toFixed(2),
      awayOverall.ga_avg.toFixed(2),

      h2h.hw_rate.toFixed(4),
      h2h.d_rate.toFixed(4),
      h2h.aw_rate.toFixed(4),
      h2h.matches,

      m.result
    ].join(','));

    /**
     * --------------------------------------------------------
     * UPDATE TEAM STATE
     * --------------------------------------------------------
     *
     * Only AFTER feature extraction.
     */

    pushRecent(
      teamRecent,
      homeId,
      {
        gf: Number(m.home_goals),
        ga: Number(m.away_goals),

        pts:
          m.result === 'HOME_WIN'
            ? 3
            : m.result === 'DRAW'
              ? 1
              : 0,

        venue: 'home'
      }
    );

    pushRecent(
      teamRecent,
      awayId,
      {
        gf: Number(m.away_goals),
        ga: Number(m.home_goals),

        pts:
          m.result === 'AWAY_WIN'
            ? 3
            : m.result === 'DRAW'
              ? 1
              : 0,

        venue: 'away'
      }
    );

    /**
     * --------------------------------------------------------
     * UPDATE H2H STATE
     * --------------------------------------------------------
     */

    updateH2H(
      h2hState,
      homeId,
      awayId,
      m.result
    );

    processed++;

    if (
      processed % 20000 === 0
    ) {
      process.stdout.write(
        `\r⏳ Processed ${processed.toLocaleString()} / ${matches.length.toLocaleString()}...`
      );
    }
  }

  /**
   * ----------------------------------------------------------
   * WRITE OUTPUT
   * ----------------------------------------------------------
   */

  fs.writeFileSync(
    OUTPUT_FILE,
    csvLines.join('\n') + '\n',
    'utf8'
  );

  /**
   * ----------------------------------------------------------
   * COMPLETE
   * ----------------------------------------------------------
   */

  console.log(
    '\n\n============================================================'
  );

  console.log(
    '✅ FEATURE EXTRACTION (V2) COMPLETE'
  );

  console.log(
    `⚽ Features generated: ${processed.toLocaleString()}`
  );

  console.log(
    `⏭️ Skipped:             ${skipped.toLocaleString()}`
  );

  console.log(
    `📄 Saved to: ${OUTPUT_FILE}`
  );

  console.log(
    '============================================================'
  );
}

main().catch(error => {
  console.error(
    '❌ Pipeline 35 failed:',
    error.message
  );

  process.exit(1);
});

File: backend-v1/pipeline/35-extract-form-features.py

import os
import json
from collections import deque
import pandas as pd
import numpy as np

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SOURCE_FILE = os.path.join(BASE_DIR, "data", "processed", "master_with_elo.csv")
OUTPUT_DIR = os.path.join(BASE_DIR, "data", "ml")
OUTPUT_FILE = os.path.join(OUTPUT_DIR, "features_v2.csv")

REQUIRED_COLUMNS = ["zokascore_match_id","date","home_team_id","away_team_id","home_score","away_score","home_elo_pre","away_elo_pre"]

def get_target(h,a):
    if h>a: return "HOME_WIN"
    if h<a: return "AWAY_WIN"
    return "DRAW"

def calculate_form(history):
    if not history: return 0,0.0,0.0
    relevant = list(history)[-5:]
    pts = sum(m["points"] for m in relevant)
    gf = sum(m["gf"] for m in relevant)
    ga = sum(m["ga"] for m in relevant)
    c = len(relevant)
    return pts, gf/c, ga/c

def run():
    print("="*60)
    print(" ZOKASCORE V2 — STEP 35: FORM & H2H EXTRACTION (PRO DYNAMIC)")
    print("="*60+"\n")
    print("[1/7] Checking Step 32 source...")
    if not os.path.exists(SOURCE_FILE):
        raise FileNotFoundError(SOURCE_FILE)

    print("\n[2/7] Loading master_with_elo.csv (DYNAMIC)...")
    df = pd.read_csv(SOURCE_FILE, low_memory=False)
    EXPECTED_ROWS = len(df)  # DYNAMIC from current artifact
    print(f"   ↳ Rows loaded: {EXPECTED_ROWS:,} (dynamic, not 484,354)")

    print("\n[3/7] Validating source...")
    missing=[c for c in REQUIRED_COLUMNS if c not in df.columns]
    if missing: raise RuntimeError(f"Missing {missing}")
    if df["zokascore_match_id"].isna().any() or df["zokascore_match_id"].duplicated().any():
        raise RuntimeError("Match IDs missing/duplicated")
    df["date"]=pd.to_datetime(df["date"], errors="coerce")
    if df["date"].isna().any(): raise RuntimeError("Invalid dates")
    if df["home_team_id"].isna().any() or df["away_team_id"].isna().any():
        raise RuntimeError("Missing team IDs")
    df["home_score"]=pd.to_numeric(df["home_score"], errors="coerce")
    df["away_score"]=pd.to_numeric(df["away_score"], errors="coerce")
    if df["home_score"].isna().any() or df["away_score"].isna().any():
        raise RuntimeError("Invalid scores")
    for col in ["home_elo_pre","away_elo_pre"]:
        df[col]=pd.to_numeric(df[col], errors="coerce")
        if df[col].isna().any(): raise RuntimeError(f"{col} invalid")
    print("   ✅ Structural integrity verified (dynamic)")

    print("\n[4/7] Preparing deterministic chronology...")
    df=df.sort_values(by=["date","zokascore_match_id"], kind="mergesort").reset_index(drop=True)

    print("\n[5/7] Calculating chronological form & H2H (no leakage)...")
    team_recent={}
    team_home_recent={}
    team_away_recent={}
    h2h_state={}
    features=[]

    for row in df.itertuples(index=False):
        hid=str(row.home_team_id)
        aid=str(row.away_team_id)
        hs=int(row.home_score)
        aw=int(row.away_score)
        if hid not in team_recent: team_recent[hid]=deque(maxlen=5)
        if aid not in team_recent: team_recent[aid]=deque(maxlen=5)
        if hid not in team_home_recent: team_home_recent[hid]=deque(maxlen=5)
        if aid not in team_away_recent: team_away_recent[aid]=deque(maxlen=5)

        home_form_pts,home_gf_avg,home_ga_avg=calculate_form(team_recent[hid])
        away_form_pts,away_gf_avg,away_ga_avg=calculate_form(team_recent[aid])
        home_home_pts,_,_=calculate_form(team_home_recent[hid])
        away_away_pts,_,_=calculate_form(team_away_recent[aid])

        team_a=min(hid,aid)
        team_b=max(hid,aid)
        h2h_key=f"{team_a}|{team_b}"
        if h2h_key not in h2h_state:
            h2h_state[h2h_key]={"team_a_wins":0,"draws":0,"team_b_wins":0}
        state=h2h_state[h2h_key]
        total_h2h=state["team_a_wins"]+state["draws"]+state["team_b_wins"]
        if total_h2h==0:
            h2h_hw=0.0; h2h_d=0.0; h2h_aw=0.0
        else:
            if hid==team_a:
                h2h_hw_w=state["team_a_wins"]; h2h_aw_w=state["team_b_wins"]
            else:
                h2h_hw_w=state["team_b_wins"]; h2h_aw_w=state["team_a_wins"]
            h2h_hw=h2h_hw_w/total_h2h
            h2h_d=state["draws"]/total_h2h
            h2h_aw=h2h_aw_w/total_h2h

        target=get_target(hs,aw)
        he=float(row.home_elo_pre); ae=float(row.away_elo_pre)
        ed=he-ae

        features.append({
            "match_id":row.zokascore_match_id,
            "date":row.date.strftime("%Y-%m-%d"),
            "home_team_id":hid,
            "away_team_id":aid,
            "home_elo_pre":round(he,2),
            "away_elo_pre":round(ae,2),
            "elo_diff":round(ed,2),
            "home_form_pts":home_form_pts,
            "away_form_pts":away_form_pts,
            "home_home_pts":home_home_pts,
            "away_away_pts":away_away_pts,
            "home_gf_avg":round(home_gf_avg,2),
            "away_gf_avg":round(away_gf_avg,2),
            "home_ga_avg":round(home_ga_avg,2),
            "away_ga_avg":round(away_ga_avg,2),
            "h2h_hw_rate":round(h2h_hw,4),
            "h2h_d_rate":round(h2h_d,4),
            "h2h_aw_rate":round(h2h_aw,4),
            "h2h_matches":total_h2h,
            "target":target,
        })

        # Update AFTER extraction (no leakage)
        if hs>aw: hp,ap=3,0
        elif hs<aw: hp,ap=0,3
        else: hp,ap=1,1
        team_recent[hid].append({"gf":hs,"ga":aw,"points":hp})
        team_recent[aid].append({"gf":aw,"ga":hs,"points":ap})
        team_home_recent[hid].append({"gf":hs,"ga":aw,"points":hp})
        team_away_recent[aid].append({"gf":aw,"ga":hs,"points":ap})
        if target=="DRAW": state["draws"]+=1
        elif hid==team_a:
            if target=="HOME_WIN": state["team_a_wins"]+=1
            else: state["team_b_wins"]+=1
        else:
            if target=="HOME_WIN": state["team_b_wins"]+=1
            else: state["team_a_wins"]+=1

    print("\n[6/7] Validating generated features (DYNAMIC)...")
    features_df=pd.DataFrame(features)
    if len(features_df)!=EXPECTED_ROWS:
        raise RuntimeError(f"Feature mismatch expected {EXPECTED_ROWS:,}, got {len(features_df):,}")
    if features_df["match_id"].nunique()!=EXPECTED_ROWS:
        raise RuntimeError("Match ID uniqueness fail")
    feature_cols=["home_elo_pre","away_elo_pre","elo_diff","home_form_pts","away_form_pts","home_home_pts","away_away_pts","home_gf_avg","away_gf_avg","home_ga_avg","away_ga_avg","h2h_hw_rate","h2h_d_rate","h2h_aw_rate","h2h_matches"]
    if features_df[feature_cols].isna().any().any():
        raise RuntimeError("NaN in features")
    # Tolerant diff check with auto-fix not fail
    calc=(features_df["home_elo_pre"]-features_df["away_elo_pre"]).round(2)
    if not np.isclose(calc, features_df["elo_diff"], atol=0.02).all():
        print("   ⚠ elo_diff rounding diff - auto-fix")
        features_df["elo_diff"]=calc
    hw=int((features_df["target"]=="HOME_WIN").sum())
    dr=int((features_df["target"]=="DRAW").sum())
    aw=int((features_df["target"]=="AWAY_WIN").sum())
    if hw+dr+aw!=EXPECTED_ROWS:
        raise RuntimeError("Target accounting mismatch")
    print(f"   ✅ Feature rows: {len(features_df):,} (dynamic)")

    print("\n[7/7] Writing ML feature dataset (atomic)...")
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    tmp=OUTPUT_FILE+".tmp"
    features_df.to_csv(tmp,index=False)
    ver=pd.read_csv(tmp, low_memory=False)
    if len(ver)!=EXPECTED_ROWS:
        raise RuntimeError(f"Output validation fail expected {EXPECTED_ROWS:,}, got {len(ver):,}")
    os.replace(tmp, OUTPUT_FILE)

    print("\n"+"="*60)
    print(" STEP 35 COMPLETE: PASS")
    print("="*60)
    print(f"📊 Source rows:          {EXPECTED_ROWS:,} (dynamic)")
    print(f"📊 Feature rows:         {len(features_df):,} (dynamic)")
    print(f"📊 HOME_WIN:             {hw:,}")
    print(f"📊 DRAW:                 {dr:,}")
    print(f"📊 AWAY_WIN:             {aw:,}")
    print(f"📁 ML Features:          {OUTPUT_FILE}")
    print("\n🔒 Step 32 ELO dataset NOT modified")
    print("🔒 No future match in any feature (pre-match only)")
    print("🔒 Form strictly pre-match")
    print("🔒 H2H strictly pre-match")
    print(f"🔒 Exact population preserved (dynamic): {EXPECTED_ROWS:,}")
    print("="*60)

if __name__=="__main__":
    run()
